pkg/pgxconn/helpers.go

```diff
@@ -0,0 +1,248 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package pgxconn
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgx/v4"
+)
+
+// msUntilCtxDeadline returns the time in milliseconds remaining
+// before the context deadline is reached.
+//
+// ok=false is returned if the given context doesn't have a deadline set.
+func msUntilCtxDeadline(ctx context.Context) (ms int64, ok bool) {
+	t, ok := ctx.Deadline()
+	if !ok {
+		return 0, false
+	}
+	return time.Until(t).Milliseconds(), true
+}
+
+// QueryWithTimeout wraps the given query in a batch along with
+// statement_timeout set to the given timeout value. The timeout value is taken
+// as milliseconds.
+//
+// The results is returned as if it was read with PgxConn.Query.
+//
+// A value of zero disables the timeout.
+//
+// The returned closeFn closes the batch operation. This must be called before
+// the underlying connection can be used again. It's safe to call multiple
+// times.
+//
+// All the queries in a batch are executed in an implicit transaction by using
+// `set local statement_timeout` the timeout it's only applied to the queries
+// belonging to the transaction and not to the connection session.
+//
+// There's a bug with the use of `set local` and implicit transactions.
+// Even though the statements are executed correctly and apply the desire
+// effects the following warning message is logged:
+//
+// `SET LOCAL can only be used in transaction blocks`
```

Comment:
This will make the DB output a lot of WARNING messages with that message. If that's a problem I can change the implementation to set an explicit transaction per batch.

It would be queueing BEGIN and COMMIT to the batch, and consuming the responses for both statements. Code-wise the downside is that using `NewBatchWithTimeoutFromCtx` and `SendBatch` won't be as pretty as we have to remember to put COMMIT at the end and to consume it.

Reply:
I think such WARNINGS would not be great and pollute the log. I would probably do an explicit transaction then. But before doing so can we do some microbenchmarks and make sure it doesn't negatively affect performance (especially in high-latency cases)

Reply:
If I read the postgres mailing list you linked correctly, it seems like prepending `SET LOCAL statement_timeout = x;` to the sql provided may have the intended effect without the warning message.

Reply:
I tried that with:

```go
r, err := conn.Query(context.Background(), "SET LOCAL statement_timeout = 4; SELECT 1, pg_sleep(3)")
```

But got the following error:

```
cannot insert multiple commands into a prepared statement (SQLSTATE 42601)
```

Reply:
Ah. Bummer.